File: app/src/main/java/info/nightscout/androidaps/data/OverlappingIntervals.java
package info.nightscout.androidaps.data;


import androidx.annotation.Nullable;

import info.nightscout.androidaps.interfaces.Interval;

/**
 * Created by adrian on 15/07/17.
 */

public class OverlappingIntervals<T extends Interval> extends Intervals<T> {

    public OverlappingIntervals() {
        super();
    }

    public OverlappingIntervals(Intervals<T> other) {
        rawData = other.rawData.clone();
    }

    protected synchronized void merge() {
        boolean needToCut = false;
        long cutTime = 0;

        for (int index = rawData.size() - 1; index >= 0; index--) { //begin with newest
            Interval cur = rawData.valueAt(index);
            if (cur.isEndingEvent()) {
                needToCut = true;
                cutTime = cur.start();
            } else {
                //event that is no EndingEvent might need to be stopped by an ending event
                if (needToCut && cur.end() > cutTime) {
                    cur.cutEndTo(cutTime);
                }
            }
        }
    }

    @Nullable
    public synchronized T getValueByInterval(long time) {
        for (int index = rawData.size() - 1; index >= 0; index--) { //begin with newest
            T cur = rawData.valueAt(index);
            if (cur.match(time)) {
                return cur;
            }
        }
        return null;
    }

}

File: core/src/main/java/info/nightscout/androidaps/data/OverlappingIntervals.kt
package info.nightscout.androidaps.data

import info.nightscout.androidaps.interfaces.Interval

class OverlappingIntervals<T : Interval> : Intervals<T> {

    constructor() : super()
    constructor(other: Intervals<T>) {
        rawData = other.rawData.clone()
    }

    @Synchronized override fun merge() {
        var needToCut = false
        var cutTime: Long = 0
        for (index in rawData.size() - 1 downTo 0) { //begin with newest
            val cur: Interval = rawData.valueAt(index)
            if (cur.isEndingEvent) {
                needToCut = true
                cutTime = cur.start()
            } else {
                //event that is no EndingEvent might need to be stopped by an ending event
                if (needToCut && cur.end() > cutTime) {
                    cur.cutEndTo(cutTime)
                }
            }
        }
    }

    @Synchronized override fun getValueByInterval(time: Long): T? {
        for (index in rawData.size() - 1 downTo 0) { //begin with newest
            val cur = rawData.valueAt(index)
            if (cur!!.match(time)) {
                return cur
            }
        }
        return null
    }
}